attest/eventlog.go

```diff
@@ -223,47 +223,36 @@ func (a *AKPublic) validate20Quote(quote Quote, pcrs []PCR, nonce []byte) error
 	}
 
 	pcrByIndex := map[int][]byte{}
-	pcrDigestLength := HashAlg(att.AttestedQuoteInfo.PCRSelection.Hash).cryptoHash().Size()
+	pcrDigestAlg := HashAlg(att.AttestedQuoteInfo.PCRSelection.Hash).cryptoHash()
 	for _, pcr := range pcrs {
-		// TODO(jsonp): Use pcr.DigestAlg once #116 is fixed.
-		if len(pcr.Digest) == pcrDigestLength {
+		if pcr.DigestAlg == pcrDigestAlg {
 			pcrByIndex[pcr.Index] = pcr.Digest
 		}
 	}
 
-	n := len(att.AttestedQuoteInfo.PCRDigest)
-	hash, ok := hashBySize[n]
-	if !ok {
-		return fmt.Errorf("quote used unsupported hash algorithm length: %d", n)
-	}
-
-	h := hash.New()
+	sigHash.Reset()
 	for _, index := range att.AttestedQuoteInfo.PCRSelection.PCRs {
 		digest, ok := pcrByIndex[index]
 		if !ok {
 			return fmt.Errorf("quote was over PCR %d which wasn't provided", index)
 		}
-		h.Write(digest)
+		sigHash.Write(digest)
 	}
 
-	if !bytes.Equal(h.Sum(nil), att.AttestedQuoteInfo.PCRDigest) {
+	if !bytes.Equal(sigHash.Sum(nil), att.AttestedQuoteInfo.PCRDigest) {
 		return fmt.Errorf("quote digest didn't match pcrs provided")
 	}
 	return nil
 }
 
-var hashBySize = map[int]crypto.Hash{
-	crypto.SHA1.Size():   crypto.SHA1,
-	crypto.SHA256.Size(): crypto.SHA256,
-}
+func extend(pcr PCR, replay []byte, e rawEvent) (pcrDigest []byte, eventDigest []byte, err error) {
+	h := pcr.DigestAlg
 
-func extend(pcr, replay []byte, e rawEvent) (pcrDigest []byte, eventDigest []byte, err error) {
-	h, ok := hashBySize[len(pcr)]
-	if !ok {
-		return nil, nil, fmt.Errorf("pcr %d was not a known hash size: %d", e.index, len(pcr))
-	}
 	for _, digest := range e.digests {
-		if len(digest) != len(pcr) {
+		if digest.hash != pcr.DigestAlg {
+			continue
+		}
+		if len(digest.data) != len(pcr.Digest) {
```

Comment:
Is the digest length check necessary given we have `if digest.hash != pcr.DigestAlg {` above? If yes, could we make it bail with an error instead of `continue` ?

Reply:
Done